src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,606 @@
+        }
+
+        progress.end();
+
+        const uint64_t bytes_used
+            = (vocab.vector.size() * sizeof(sgns_vocab_vector::value_type))
+              + vocab.table.bytes_used();
+
+        LOG(info) << "Loaded vocabulary of size " << vocab.table.size()
+                  << " occupying " << printing::bytes_to_units(bytes_used)
+                  << ENDLG;
+
+        if ((max_ram_ <= bytes_used)
+            || ((max_ram_ - bytes_used) < (1024 * 1024)))
+        {
+            throw sgns_exception{"RAM limit too restrictive"};
+        }
+
+        return vocab;
+    }
+
+    // Create a vector whose elements are present in proportion
+    // to their noise sample probablility
+    sgns_noise_distribution
+    create_unigram_noise_distribution(sgns_noise_distribution::size_type size,
+                                      double power) const
+    {
+        printing::progress progress{" > Generating noise distribution: ",
+                                    vocab_.vector.size()};
+
+        sgns_noise_distribution noise_dist(size);
+
+        double normalizer = 0;
+        for (sgns_vocab_vector::size_type i = 0; i < vocab_.vector.size(); ++i)
+            normalizer += std::pow(vocab_.vector[i].count, power);
+
+        sgns_noise_distribution::size_type i = 0;
+        double cumulative_probability = 0;
+
+        for (sgns_vocab_vector::size_type j = 0; j < vocab_.vector.size(); ++j)
+        {
+            progress(j);
+            cumulative_probability
+                += std::pow(vocab_.vector[j].count, power) / normalizer;
+
+            while ((i < noise_dist.size())
+                   && ((float)i / noise_dist.size() < cumulative_probability))
+            {
+                noise_dist[i] = j;
+                ++i;
+            }
+        }
+
+        progress.end();
+
+        LOG(info) << "Created smoothed unigram noise distribution of size "
+                  << noise_dist.size() << ENDLG;
+
+        return noise_dist;
+    }
+
+    // Save vectors in MeTA format.
+    void save_meta_vectors() const
+    {
+        // Target embeddings.
+        {
+            std::ofstream output{prefix_ + "/embeddings.target.bin",
+                                 std::ios::binary};
+            printing::progress progress{" > Saving target embeddings: ",
+                                        net_.syn0.size()
+                                            + vocab_.vector.size()};
+            uint64_t words_processed = 0;
+
+            io::packed::write(output, vector_size_);
+
+            // Save target vectors.
+            for (auto i : net_.syn0)
+            {
+                io::packed::write(output, (double)i);
+                progress(++words_processed);
+            }
+
+            // Write out the unk vector.
+            for (std::size_t i = 0; i < vector_size_; ++i)
+            {
+                io::packed::write(output, 0.0);
+                progress(++words_processed);
+            }
```

Comment:
Nope. At the time I wasn't sure if it would make an impact or not. I can add it if you think it will provide an improvement.